### PR TITLE
Some (hopefully) helpful and definitely well-meaning commentary

### PR DESCRIPTION
**Pull request by christianromney**
No need to actually pull, Pull Requests are just easy communication tools...

### CODE REVIEW

**Comment by bradland** [comment]
The comments about personal growth were endearing, but the whole bit about "you shouldn't reimplement sed or grep" is really bad advice, IMO. The fact that a great tool already exists is about the worst reason for _not_ writing software that I can think of. Neither sed, nor grep, nor awk were as fast or powerful as they are today when they were first written. This tool obviously scratched an itch for the author. One person's opinion cannot invalidate the possibility that others will find it similarly useful.

**Comment by christianromney** [comment]
I agree with itch scratching whole-heartedly. I view that as primarily educational or entertaining. As for reimplementing existing tools, I think it's fine to specialize or generalize programs to fit different needs and use cases. Some of these efforts will bear fruit, others won't. But looking at these specific implementations they don't really address grep or sed's shortcomings or add missing functionality. That doesn't mean the exercise was futile, but one has to be clear about the goals.

**Comment by christianromney** [comment]
@bradland I also went back and re-read my comments. Here's a small excerpt, but I think you may find on closer inspection that you and I are probably saying the same thing in different ways:

// The first thing that strikes me is that 'search' and 'replace' reinvent
// functionality that is found in existing utilities. This is not always
// a bad thing. I can think of several great reasons to do so:
//
// - you believe the existing program to be of poor quality (for any definition of quality)
// - you wish to specialize (or generalize) the behavior of a program
// - you wish to 'port' a program to an unsupported platform
// - you with to undertake the task of rewriting a program as a learning exercise
//
// In this case, the utilities we are primarily considering are 'grep' and 'sed'. 
// There is certainly a case to be made that grep's "user interface" is not suited 
// to novices. In fact, the popular code search tool 'ack' styles itself as a better
// grep for code search.
//
// On the other hand, grep and sed are ubiquitous. They can be found on every Unix/Linux box,
// and Mac. They are also incredibly powerful utilities offering much more functionality and
// likely better performance than these scripts. When searching a large codebase, speed matters.
// I submit to you that no matter how inhumane their interfaces may seem at first glance that it is
// worthwhile to learn to wield them well. Your effort will be repaid many times over in both 
// increased productivity and peer recognition. I also believe learning these tools will help you
// understand other tools (and even software architectures!) better.

**Comment by petdance** [comment]
_I submit to you that no matter how inhumane their interfaces may seem at first glance that it is worthwhile to learn to wield them well._

Who says she doesn't know how to wield them well?  Just because she can doesn't mean she wants to.  Why should everyone perform a task using identical tools?

_Your effort will be repaid many times over in both increased productivity and peer recognition._

Your assumption that `sed` makes her more productive than `replace` has no basis.

As to peer recognition, I appreciate that she created a tool that solved a need she found and published it for others to use.  That's what open source is all about.

_the popular code search tool 'ack' styles itself as a better grep for code search._

Not only ack.  There are [at least a dozen tools that are better than grep](http://betterthangrep.com/more-tools/) for searching source code.  I'd ask that you embrace them rather than scorning them because they don't scratch your particular itch.

**Comment by bradland** [comment]
There's a comment on HackerNews that intends to explain why this tool isn't justified, but in my view, does a really great job of illustrating why I actually prefer it.

The examples provided by the author are:

```
replace 'var' 'let' *
sed -ri 's,var,let,g' *

replace 'var' 'let' . -r
find . -type f -print0 | xargs -0 sed -ri 's,var,let,g'

replace 'var' 'let' test/file1.js test/file2.js
sed -ri 's,var,let,g' test/file1.js test/file2.js

replace '(\w+)_(\w+)' '$1-$2' *
sed -ri 's,(\w+)_(\w+),\1-\2,g' *

replace 'var' 'let' . -r --include="*.js"
find . -type f -name '*.js' -print0 | xargs -0 sed -ri 's,var,let,g'

replace 'var' 'let' . -r --exclude="*.min.js,*.py"
find . -type f ! -name '*.min.js' ! -name '*.py' -print0 | xargs -0 sed -ri 's,var,let,g'

replace 'var' 'let' . -r --preview
find . -type f -exec sh -c "echo == {}; sed -r 's,var,let,g' {}" \;
(not sure if --preview does exactly that, I echo the filename followed by the modified content)
```

Source: http://news.ycombinator.com/item?id=5107491

**Comment by christianromney** [comment]
@petdance 

"Who says she doesn't know how to wield them well?"

It is a pre-emptive argument, and I may be indeed be wrong, but the context of the post led me to believe this is the case. In either case, I consider it to be good general advice. Anyone who takes the time to learn the core utils will be well served. I really don't think this is controversial. 

"Just because she can doesn't mean she wants to."

Irrefutable, if unlikely.

"Why should everyone perform a task using identical tools?"

Strawman argument. Please feel free to use whatever tool you like. However, the fact that I currently prefer to use Emacs does not in any way detract from the argument that I would be well served to learn how to use Vim, too.

"Your assumption that sed makes her more productive than replace has no basis."

I disagree with your assertion that learning a tool which is ubiquitous and more powerful doesn't make one more productive than using a less capable approximation of the same tool which is only found on a handful of machines and which requires node.js as pre-requisite. 

"As to peer recognition, I appreciate that she created a tool that solved a need she found and published it for others to use. That's what open source is all about."

These are not mutually exclusive. I can also appreciate that. In fact, I encouraged it in my closing remarks. I nevertheless admire the superior familiarity that many of my colleagues have with powerful and Unix command line tools. 

"Not only ack. There are at least a dozen tools that are better than grep for searching source code. I'd ask that you embrace them rather than scorning them because they don't scratch your particular itch." 

You must be reading something other than what I wrote. At no point did I scorn any tool. At no point did I state that Ack is the only alternative to grep. I may be reading too much into this, but I detect some misplaced outrage... In fact, I find it ironic that you are advocating that I embrace ack and its ilk, while simultaneously refuting my advice to embrace existing tools.

**Comment by petdance** [comment]
> you are advocating that I embrace ack and its ilk

Let me rephrase.  I don't ask that you embrace ack and other tools like it themselves.  I ask that you embrace their existence as a good thing.

**Comment by christianromney** [comment]
@bradland Regarding the usage examples from HN: I agree that the user interface or "syntax" of replace's commands are more humane, nicer, or easier to remember—at least for these cases. How would one pipe an existing file to replace and delete occurrences of the word 'hello'? How quickly, compared to sed, could it replace 'big' with 'huge' and 'good' with 'great' across thousands of large text files? Could it do so in one pass? How likely are you to find replace on your ISP's web server? 

Mind you, I am completely aware that these may not have been design goals for 'replace'. And without the benefit of the design rationale, it's difficult to judge. But at this point the debate is getting academic. The intent of my post was merely to offer some possible insight into some of the questions Heather posed in her post. She seemed to be genuinely interested in constructive criticism. I also mentioned the use of eval and subroutines, for instance which may be more to the point. In any case, I hope she will choose to read my comments in the spirit in which they were offered. I also hope she will consider whether or not I am full of shit and make up her own mind.

**Comment by christianromney** [comment]
@petdance Fair enough. For what it's worth, I agree. Having options can be wonderfully valuable. Cheers.

**Comment by bradland** [comment]
> She seemed to be genuinely interested in constructive criticism.

@xmlblog The problem is that I see more criticism than constructivism. I'm not sure what the pragmatic outcome of your advice should be. We have no other choice but to make the discussion academic, because your advice was short on details when it came to actual software design advice and long on personal growth advice.

I don't doubt that you had good intentions, but what was the goal? To convey that this software should never have been written? It's difficult to engage in a debate over how someone else should spend their time when they're not asking you. I read her posts, and I didn't see many questions, so maybe I'm missing something, but by my read, she was wondering what was wrong with her code.

**Comment by christianromney** [comment]
@bradland You make some good points. The only pragmatic outcome of my advice is reflection. I perceived that Heather had asked, "what's wrong with my code?"

While speculative, I touched on a few of those points. Lack of subroutines, eval, duplication. These are implementation concerns. At least one of the Twitter comments questioned whether or not writing this utility was a good idea in and of itself. For me, that question is impossible to answer without knowing the author's design goals or intent. I listed several reasons why writing this utility might be a really good idea. Perhaps I should have spent as much time elaborating those as I did highlighting a couple of reasons why it might have been perceived to be a bad idea. 

However, I felt the author already had her own compelling motivation for writing the software since she actually wrote it and it exists. This is an analog to the anthropic principle. What she wanted to understand was why someone might feel this effort was either a waste of her time or a bad idea. Without asserting that either of those is actually true, I attempted to supply some possible answers to that question. 

For what it's worth, had this been my code I would feel that it was definitely worth writing if for no other reason than the discussion it has provoked.
